### PR TITLE
Prevent failures in the "scanning for endstream" code, in `Parser_makeStream`, by handling the case where 'endstream' is split between contiguous chunks (issue 1536)

### LINKED ISSUE #1536: Missing endstream error when loading a pdf
https://w2.eff.org/legal/cases/betamax/betamax_oral_argument2.pdf

Below is the information provided by pdf.js

PDF.JS Build: 7c8f5d4
Message: Missing endstream
Stack: error("Missing endstream")@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:927
Parser_makeStream([object Object],(void 0))@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:25771
Parser_getObj()@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:25658
XRef_fetch([object Object])@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:3288
XRef_fetchIfRef([object Object])@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:3240
Page_getOperatorList([object Object],[object Array])@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:226
wphSetupPageRequest(3)@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:28932
messageHandlerComObjOnMessage([object MessageEvent])@blob:11f5bf5e-1b6d-4aae-81eb-ca5c5da84c00:28866

Nightly 14.0a1 (2012-04-12)

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Fixes #1536.

This patch also includes a test-case for issue #2098. That was fixed by PR #3376, which didn't include a test, so it seemed like a good idea to improve the test-coverage for this part of the codebase.

### CODE REVIEW

File: src/core/parser.js [hunk 1]
```diff
@@ -507,8 +507,12 @@ var Parser = (function ParserClosure() {
...
             break;
           }
-          skipped += scanLength;
-          stream.pos += scanLength;
+          // Since the stream is scanned in `SCAN_BLOCK_SIZE` sized chunks,
+          // it's thus possible that an 'endstream' might be split between
+          // two contiguous `scanBytes` chunks. Avoid skipping over a valid
+          // 'endstream' by subtracting `j` here (fixes issue1536.pdf).
+          skipped += scanLength - j;
+          stream.pos += scanLength - j;
```
Comment (yurydelendik) [inline src/core/parser.js]: We already doing subtraction at line 486 above. Which shall properly protect from the case you described in the comment above. Interesting what's missing something the above?

Comment (yurydelendik) [inline src/core/parser.js]: Oh, I see it's just bad loop structure above. We shall re-write it as:

  while(i < scanLength) {
    j = 0;
    while (j < ENDSTREAM_SIGNATURE_LENGTH && scanBytes[i + j] === ENDSTREAM_SIGNATURE[j]) j++;
    if (j >= ENDSTREAM_SIGNATURE_LENGTH) {
                 found = true;
                 break;
    }
    i++;
  }

which easier-to-read equivalent of the code above without error.

Comment (Snuffleupagus) [inline src/core/parser.js]: Nice, thanks for the help here! I've pushed a new version that addresses this.

**Comment by yurydelendik** [comment]
Thank you for the patch

/botio makeref

**Pull request merged**